src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+
+            // Acquire credentials
+
+            var credentialsUri = new Uri(baseUri, "latest/meta-data/iam/security-credentials/");
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(credentialsUri, roleName),
+                Method = HttpMethod.Get
+            };
+            credentialsRequest.Headers.Add("X-aws-ec2-metadata-token", token);
+            var credentials = await GetHttpContent(credentialsRequest, "Failed to acquire EC2 credentials.").ConfigureAwait(false);
+
+            return credentials;
+        }
+
+        private static async Task<string> GetECSResponse(Uri baseUri, string relativeUri)
+        {
+            var credentialsRequest = new HttpRequestMessage
+            {
+                RequestUri = new Uri(baseUri, relativeUri),
+                Method = HttpMethod.Get
+            };
+
+            return await GetHttpContent(credentialsRequest, "Failed to acquire ECS credentials.").ConfigureAwait(false);
+        }
+
+        private static async Task<string> GetHttpContent(HttpRequestMessage request, string exceptionMessage)
+        {
+            HttpResponseMessage response;
+            try
+            {
+                response = await _httpClientInstance.Value.SendAsync(request).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            catch (HttpRequestException ex)
+            {
+                throw new MongoClientException(exceptionMessage, ex);
+            }
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new MongoClientException(exceptionMessage);
+            }
+
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        }
+
+        private static void ValidateCredentials(string username, string password, string sessionToken)
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        private static void ValidateCredentials(string username, SecureString password, string sessionToken)
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
```

Comment:
Regarding validation please refer to spec [(link)](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#obtaining-credentials):

> regardless of how Drivers obtain credentials the only valid combination of credentials is an access key ID and a secret access key or an access key ID, a secret access key and a session token.